cpp/src/arrow/table.cc

```diff
@@ -532,6 +550,19 @@ bool Table::Equals(const Table& other) const {
   return true;
 }
 
+Status Table::Compact(MemoryPool* pool, std::shared_ptr<Table>* out) const {
+  const int ncolumns = num_columns();
+  std::vector<std::shared_ptr<Column>> compacted_columns;
+  compacted_columns.reserve(ncolumns);
+  for (int i = 0; i < ncolumns; ++i) {
+    std::shared_ptr<Column> compacted_column;
```

Comment:
You don't need this temporary, just pass `&compated_columns[i]` to the next function, and remove the line after.